File: adaptive_repair/data/python_programs/possible_change.py
def possible_change(coins, total):
    # Initialize a DP table where dp[i] will store the number of ways to make change for amount i.
    # The size of the table is total + 1 to account for amounts from 0 to total.
    dp = [0] * (total + 1)

    # There is one way to make change for 0 (by using no coins).
    dp[0] = 1

    # Iterate through each coin denomination.
    # The order of loops (coins first, then amounts) is crucial for counting combinations
    # (where the order of coins used doesn't matter).
    for coin in coins:
        # For each coin, iterate from the coin's value up to the total amount.
        # This ensures that we consider using the current coin for each possible amount.
        for amount in range(coin, total + 1):
            # The number of ways to make 'amount' is the sum of:
            # 1. The ways to make 'amount' without using the current 'coin' (which is dp[amount] before this update).
            # 2. The ways to make 'amount - coin' using any of the coins considered so far (including the current 'coin'),
            #    and then adding the current 'coin' to those combinations.
            dp[amount] += dp[amount - coin]

    # The final result is the number of ways to make change for the 'total' amount.
    return dp[total]

File: adaptive_repair/data/java_programs/POSSIBLE_CHANGE.java
package java_programs;

public class POSSIBLE_CHANGE {
    public static int possible_change(int[] coins, int total) {
        // Handle edge case for negative total
        if (total < 0) {
            return 0;
        }

        // dp[i] will store the number of ways to make amount i
        // Initialize dp array with zeros, size total + 1
        int[] dp = new int[total + 1];

        // Base case: There is one way to make amount 0 (by using no coins)
        dp[0] = 1;

        // Iterate through each coin denomination
        for (int coin : coins) {
            // For each coin, update the dp array for amounts from 'coin' up to 'total'
            // This ensures that each coin is considered for all possible amounts it can contribute to.
            for (int i = coin; i <= total; i++) {
                // Add the number of ways to make the remaining amount (i - coin)
                // This represents using the current 'coin' to reach amount 'i'.
                dp[i] += dp[i - coin];
            }
        }

        // The final result is the number of ways to make the 'total' amount
        return dp[total];
    }
}